src/kmeans.jl

```diff
@@ -1,91 +1,125 @@
 # K-means algorithm
 
-#### Interface
-
-mutable struct KmeansResult{T<:Real} <: ClusteringResult
-    centers::Matrix{T}         # cluster centers (d x k)
-    assignments::Vector{Int}   # assignments (n)
-    costs::Vector{T}           # costs of the resultant assignments (n)
-    counts::Vector{Int}        # number of samples assigned to each cluster (k)
-    cweights::Vector{Float64}  # cluster weights (k)
-    totalcost::Float64         # total cost (i.e. objective) (k)
-    iterations::Int            # number of elapsed iterations
-    converged::Bool            # whether the procedure converged
+####
+#### Result object
+####
+
+struct KmeansResult{TC<:AbstractFloat,TD<:Real,TCW<:Real} <: ClusteringResult
+    centers::Matrix{TC}       # cluster centers (d x k)
+    assignments::Vector{Int}  # assignments (n)
+    costs::Vector{TD}         # cost of the assignments (n)
+    counts::Vector{Int}       # number of samples assigned to each cluster (k)
+    cweights::Vector{TCW}     # cluster weights (k)
```

Comment:
I'm not sure 100% we need so many type parameters, maybe `Vector{<:Real}` or `Vector{Float64}` would do just fine.
@nalimilan ?
Anyway, I'd use `T` instead of `TC`, `C` instead of `TD`, `W` instead of `TCW`.

Reply:
So the reasoning is that now the code in the current PR has no type casting (no loss of precision) and `costs` and `cweights` are defined by the call:

* `costs` is effectively only defined via `dmat` which is defined as the distance between TX and TC elements --> so it's a parameter, it could be encapsulated by a Float64 but I don't think that's necessary/makes more sense
* `cweights` is effectively only defined by `weights`, if `weights==nothing` then it's a vector of Int, otherwise it's a vector of type `eltype(weights)` --> again a parameter

naming of parameters I really don't care so whatever floats your boat. The advantage with these two letter things is that their use is consistent throughout the code and so possibly less confusing. 🤷‍♂️

Reply:
I agree single-letter names are enough. Maybe we could drop at least the parameter for `costs`, and use either the same eltype as `centers`? Since we allocate the matrix manually, it shouldn't be an issue. Or just use `Float64`, which is what `dmat` will use AFAICT since it results from `pairwise`.

Reply:
I don't agree with this, `costs` is exclusively tied to `dmat` which is not guaranteed be of type `centers` and the conversion that existed before to guarantee this is unnecessary IMO.

And re single letter names, it's fine, however there's a different type for `weights` an `clusterweights` and for `centers` and `costs` so that's going to be something like `W`, `E`, `C`, `K` which ends up being less clear imo.

Reply:
`cweights` is never `nothing`. It's either `Int` (unweighted case) or `eltype(weights)` provided `weights != nothing`.

The use-case is just that if all computations are done in `Float32` it seems silly to widen the type to `Float64`. I actually think the current type parametrization is less error prone than the previous.

Ps: I did add tests for these types.

Reply:
> `cweights` is never `nothing`. It's either `Int` (unweighted case) or `eltype(weights)` provided `weights != nothin`

But for the weightless case shouldn't the `counts` be identical to `cweights`? For me it seems reasonable that if the user did not provide `weights` then `cweights` are also `nothing`.